### PR TITLE
Feature 105 set values of missing ec_value fields

### LINKED ISSUE #105: Add processing to set value of field ec_value for CTC, CTS, MCTC, and MCTS records
*If ec_value is missing for CTC and CTS records, set the value to .5. If ec_value is missing for MCTC and MCTS records, set the value to be 1/n_cat.*

### Acceptance Testing ###
*Load records of all 4 types for testing*

### Time Estimate ###
*Estimate the amount of work required here.*
*1 day of work.*

### Sub-Issues ###
Consider breaking the new feature down into sub-issues.
- [ ] *Add a checkbox for each sub-issue here.*

### Relevant Deadlines ###
*List relevant project deadlines here or state NONE.*

### Funding Source ###
*Define the source of funding and account keys here or state NONE.*

### Assignee ###
- [ ] Select **engineer(s)** or **no engineer** required
- [ ] Select **scientist(s)** or **no scientist** required

### Labels ###
- [ ] Select **component(s)**
- [ ] Select **priority**
- [ ] Select **requestor(s)**

### Projects and Milestone ###
- [ ] Select **Repository** and/or **Organization** level **Project(s)** or add **alert: NEED PROJECT ASSIGNMENT** label
- [ ] Select **Milestone** as the next official version or **Future Versions**

## Define Related Issue(s) ##
Consider the impact to the other METplus components.
- [ ] [METplus](https://github.com/dtcenter/METplus/issues/new/choose), [MET](https://github.com/dtcenter/MET/issues/new/choose), [METdataio](https://github.com/dtcenter/METdataio/issues/new/choose), [METviewer](https://github.com/dtcenter/METviewer/issues/new/choose), [METexpress](https://github.com/dtcenter/METexpress/issues/new/choose), [METcalcpy](https://github.com/dtcenter/METcalcpy/issues/new/choose), [METplotpy](https://github.com/dtcenter/METplotpy/issues/new/choose)

## New Feature Checklist ##
See the [METplus Workflow](https://metplus.readthedocs.io/en/latest/Contributors_Guide/github_workflow.html) for details.
- [ ] Complete the issue definition above, including the **Time Estimate** and **Funding source**.
- [ ] Fork this repository or create a branch of **develop**.
Branch name: `feature_<Issue Number>_<Description>`
- [ ] Complete the development and test your changes.
- [ ] Add/update log messages for easier debugging.
- [ ] Add/update unit tests.
- [ ] Add/update documentation.
- [ ] Push local changes to GitHub.
- [ ] Submit a pull request to merge into **develop**.
Pull request: `feature <Issue Number> <Description>`
- [ ] Define the pull request metadata, as permissions allow.
Select: **Reviewer(s)** and **Development** issues
Select: **Repository** level development cycle **Project** for the next official release
Select: **Milestone** as the next official version
- [ ] Iterate until the reviewer(s) accept and merge your changes.
- [ ] Delete your fork or branch.
- [ ] Close this issue.

### PR DESCRIPTION
**Pull request by venitahagerty**
## Pull Request Testing ##

Made changes per https://github.com/dtcenter/METdataio/issues/105

- [ ] Describe testing already performed for these changes:</br>
Ran the MET test data. New values for ec_value appear correct

- [ ] Recommend testing for the reviewer(s) to perform, including the location of input datasets, and any additional instructions:</br>

- [ ] Do these changes include sufficient documentation updates, ensuring that no errors or warnings exist in the build of the documentation? **[Yes or No]**

- [ ] Do these changes include sufficient testing updates? **[Yes or No]**

- [ ] Will this PR result in changes to the test suite? **[Yes or No]**</br>
If **yes**, describe the new output and/or changes to the existing output:</br>

- [ ] Please complete this pull request review by **[Fill in date]**.</br>

## Pull Request Checklist ##
See the [METplus Workflow](https://metplus.readthedocs.io/en/latest/Contributors_Guide/github_workflow.html) for details.
- [ ] Review the source issue metadata (required labels, projects, and milestone).
- [ ] Complete the PR definition above.
- [ ] Ensure the PR title matches the feature or bugfix branch name.
- [ ] Define the PR metadata, as permissions allow.
Select: **Reviewer(s)**
Select: **Organization** level software support **Project** or **Repository** level development cycle **Project**
Select: **Milestone** as the version that will include these changes
- [ ] After submitting the PR, select **Development** issue with the original issue number.
- [ ] After the PR is approved, merge your changes. If permissions do not allow this, request that the reviewer do the merge.
- [ ] Close the linked issue and delete your feature or bugfix branch from GitHub.

### CODE REVIEW

**Comment by venitahagerty** [comment]
I checked to see if changing ubuntu-18.04 to ubuntu-latest would be a problem, and it is. I'll create an issue for that change.

**Comment by TatianaBurek** [comment]
Should the default values for ec_value be set to 0.5 in mv_mysql.sql?

**Comment by venitahagerty** [comment]
Changed mv_mysq.sql to have the default value for ec_value in CTC/CTS be .5. ec_value is used in METviewer as a filter so it can't be -9999.

File: METdbLoad/ush/read_data_files.py [hunk 1]
```diff
@@ -630,6 +630,23 @@ def read_data(self, load_flags, load_files, line_types):
                                          (all_stat['8'].isnull()) &
                                          (~all_stat['5'].isnull()), '5']
 
+                # Some lines in stat files may be missing ec_value
+                # CTC and CTS, set to .5
+                # MCTS, set to 1/n_cat. MCTC is variable length
+                if all_stat[CN.LINE_TYPE].eq(CN.CTC).any():
+                    all_stat.loc[(all_stat.line_type == CN.CTC) &
+                                 (all_stat['5'].isnull()), '5'] = .5
```
Comment (TatianaBurek) [inline METdbLoad/ush/read_data_files.py]: Usually the missing value in MET output has value 'NA'. We need to add a check against NA here:
if (CTS or CTS) and (all_stat['5'] is null or NA)

File: METdbLoad/ush/read_data_files.py [hunk 2]
```diff
@@ -630,6 +630,23 @@ def read_data(self, load_flags, load_files, line_types):
...
+
+                if all_stat[CN.LINE_TYPE].eq(CN.CTS).any():
+                    all_stat.loc[(all_stat.line_type == CN.CTS) &
+                                 (all_stat['96'].isnull()), '96'] = .5
```
Comment (TatianaBurek) [inline METdbLoad/ush/read_data_files.py]: See comment above

Comment (venitahagerty) [reply METdbLoad/ush/read_data_files.py]: I checked the Python docs to see if both isna and isnull are needed in pandas. The definitions are the same, and an entry in Stack Overflow says that they are literally the same code. They both check for null or missing values.

**Comment by TatianaBurek** [comment]
I tested by changing a CTC row with ec_value=0.5 to NA and sow that the script did not switch the value to 0.5 and saved it as -9999
I suspect that Python reads NA from the file as string so isna would not work and  it might be useful to add
 all_stat['5'] == 'NA'

**Comment by venitahagerty** [comment]
Added checks for METviewer Not Available (CN.NOTAV = 'NA'). Also handled VSDB FHO lines which are converted to CTC lines.

**Review by TatianaBurek — APPROVED** [review]

**Pull request merged**